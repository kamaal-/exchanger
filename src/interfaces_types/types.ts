import Rates from "./Rates";
import Wallet from "./Wallet";
import Transaction from "./Transaction";

export type NullableRates = Rates | null;
export type NullableString = string | null;
export type NullableNumber = number | null;
export type NullableRate = Rate | null;
export type NullableWallets = Wallet[] | null;
export type NullableWallet = Wallet | null;
export type NullableBalance = Balance | null;

export type PairWallets = {
  fromWallet: Wallet;
  toWallet: Wallet;
};

export type Balance = {
  currency: string;
  balance: number;
  diff: number;
  label: string;
};

export type BalancePair = {
  [s: string]: Balance;
};

export type StoreState = {
  Rates: NullableRates;
  fromRate: Rate | null;
  toRate: Rate | null;
  error: NullableString;
  Wallets: WalletsState;
};

export type RatesState = {
  rates: NullableRates;
  fromRate: Rate | null;
  toRate: Rate | null;
  error: NullableString;
  base: NullableString;
};

export type WalletsState = {
  wallets: NullableWallets;
  fromWallet: NullableWallet;
  toWallet: NullableWallet;
  balances: BalancePair | null;
  leftWallet: NullableWallet;
  rightWallet: NullableWallet;
  transactions: Transaction[];
  value: string;
};

export type Action = {
  type: string;
  payload: any;
};

export type ConverterArgs = {
  fromType: string;
  toType: string;
  value: number;
  rates: Rates;
  base?: string;
  decimalPlaces?: number;
  fallback?: number;
};

export type Rate = {
  currency: string;
  rate: number;
};

export enum Size {
  Large = "Large",
  Medium = "Medium",
  Small = "Small"
}

export enum Align {
  Left,
  Right
}

export enum ButtonTypes {
  Main,
  Secondary,
  Disabled
}

export enum Theme {
  Dark,
  Light
}

export enum WalletType {
  From = "fromWallet",
  To = "toWallet"
}

export enum WalletSides {
  Left = "leftWallet",
  Right = "rightWallet"
}

// Props

export type InputProps = {
  wallet: Wallet;
  exchangeWallet: Wallet;
  theme: Theme;
  triggerConversion?: (
    input: string,
    fromWallet: Wallet,
    toWallet: Wallet
  ) => void;
  setFromWallet?: (w: Wallet) => void;
  setToWallet?: (w: Wallet) => void;
};

export type WalletProps = {
  wallet: Wallet;
  exchangeWallet: Wallet;
  theme: Theme;
  type: WalletType;
  balances: BalancePair;
  toWallet: Wallet;
  exchange: () => void;
};
