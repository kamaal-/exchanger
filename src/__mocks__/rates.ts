export const rates = {
    "EUR": {
        currency: 'EUR',
        rate: 0.9096697899
    },
    "SEK": {
        currency: 'SEK',
        rate: 9.9072136814
    },
    "USD": {
        currency: 'USD',
        rate: 1.0
    },
    "GBP": {
        currency: 'GBP',
        rate: 0.8110161012
    }
};

export const rate2 = {
    "EUR": {
        currency: 'EUR',
        rate: 1.9096697899
    },
    "SEK": {
        currency: 'SEK',
        rate: 9.9072136814
    },
    "USD": {
        currency: 'USD',
        rate: 1.0
    },
    "GBP": {
        currency: 'GBP',
        rate: 0.8110161012
    }
};