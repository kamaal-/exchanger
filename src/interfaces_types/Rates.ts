import {Rate} from './types'
export default interface Rates{ 
    [s: string]: Rate; 
}