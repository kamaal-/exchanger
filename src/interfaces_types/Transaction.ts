
export default interface Transaction{
    detail:string;
    wallets: {
        from: {
            currency: string;
            symbol: string;
        };
        to: {
            currency: string;
            symbol: string;
        };
    };
    from:{
        before: string;
        diff: string;
        after: string;
    }
    to: {
        before: string;
        diff: string;
        after: string;
    }

    exchange: string;
    time: number;
}