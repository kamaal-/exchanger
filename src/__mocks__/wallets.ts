import IWallet from "../interfaces_types/Wallet";
import { rates } from "./rates";
import Wallet from "../services/models/Wallet";
import { symbol } from "prop-types";

export const wallets: IWallet[] = [
  {
    currency: rates["EUR"].currency,
    rate: rates["EUR"].rate,
    balance: 10.23,
    symbol: "€"
  },
  {
    currency: rates["USD"].currency,
    rate: rates["USD"].rate,
    balance: 150.45,
    symbol: "$"
  },
  {
    currency: rates["GBP"].currency,
    rate: rates["GBP"].rate,
    balance: 320.4,
    symbol: "l"
  }
];

export const balances = {
  fromWallet: {
    currency: "EUR",
    balance: 10.23,
    diff: 0,
    label: "Balance"
  },
  toWallet: {
    currency: "USD",
    balance: 150.45,
    diff: 0,
    label: "Balance"
  }
};

export const balances_ok = {
  fromWallet: {
    currency: "EUR",
    balance: 10.23,
    diff: -3.23,
    label: "Balance"
  },
  toWallet: {
    currency: "USD",
    balance: 150.45,
    diff: 4.56,
    label: "Balance"
  }
};

export const balances_not_ok = {
  fromWallet: {
    currency: "EUR",
    balance: 10.23,
    diff: -10.24,
    label: "Balance"
  },
  toWallet: {
    currency: "USD",
    balance: 150.45,
    diff: 4.56,
    label: "Balance"
  }
};


export const transactions = [ 
  { 
     "detail":"You have transfered 34 kr to RM.",
     "wallets":{ 
        "from":{ 
           "currency":"SEK",
           "symbol":"kr"
        },
        "to":{ 
           "currency":"MYR",
           "symbol":"RM"
        }
     },
     "from":{ 
        "before":"100 kr",
        "diff":"34 kr",
        "after":"66 kr"
     },
     "to":{ 
        "before":"200 RM",
        "diff":"14.47 RM",
        "after":"214.47 RM"
     },
     "exchange":"1 kr = 0.42 RM",
     "time":1570979215945
  },
  { 
     "detail":"You have transfered 2 € to £.",
     "wallets":{ 
        "from":{ 
           "currency":"EUR",
           "symbol":"€"
        },
        "to":{ 
           "currency":"GBP",
           "symbol":"£"
        }
     },
     "from":{ 
        "before":"95 €",
        "diff":"2 €",
        "after":"93 €"
     },
     "to":{ 
        "before":"104.38 £",
        "diff":"1.75 £",
        "after":"106.13 £"
     },
     "exchange":"1 € = 0.87 £",
     "time":1570979187093
  },
  { 
     "detail":"You have transfered 5 € to £.",
     "wallets":{ 
        "from":{ 
           "currency":"EUR",
           "symbol":"€"
        },
        "to":{ 
           "currency":"GBP",
           "symbol":"£"
        }
     },
     "from":{ 
        "before":"100 €",
        "diff":"5 €",
        "after":"95 €"
     },
     "to":{ 
        "before":"100 £",
        "diff":"4.38 £",
        "after":"104.38 £"
     },
     "exchange":"1 € = 0.87 £",
     "time":1570973116868
  }
];