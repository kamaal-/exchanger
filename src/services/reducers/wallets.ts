import { WalletsState, Action } from "../../interfaces_types/types";
import { WALLETS, CONVERT, APP } from "../action_types/index";

export const initialState: WalletsState = {
  wallets: null,
  fromWallet: null,
  toWallet: null,
  balances: null,
  leftWallet: null,
  rightWallet: null,
  transactions: [],
  value: ""
};

export default (state = initialState, { type, payload }: Action) => {
  switch (type) {
    case WALLETS.ADD:
      return { ...state, wallets: payload };
    case WALLETS.REMOVE:
      return { ...state, wallets: payload };
    case WALLETS.FROM:
      return { ...state, fromWallet: payload };
    case WALLETS.TO:
      return { ...state, toWallet: payload };
    case WALLETS.SWITCH:
      return {
        ...state,
        toWallet: state.fromWallet,
        fromWallet: state.toWallet
      };
    case CONVERT.DO:
      return { ...state };
    case WALLETS.LEFT:
      return { ...state, leftWallet: payload };
    case WALLETS.RIGHT:
      return { ...state, rightWallet: payload };
    case WALLETS.RESET:
      return {
        ...state,
        balances: payload.balances,
        fromWallet: payload.fromWallet,
        toWallet: payload.toWallet
      };
    case CONVERT.RUN:
      return { ...state, balances: payload };
    case WALLETS.ADD_TRANSACTION:
      return { ...state, transactions: payload };
    case WALLETS.SET_VALUE:
      return { ...state, value: payload };
    case APP.SUCCESS_POLLING:
    case APP.START_POLLING:
      return { ...state };
    default:
      return initialState;
  }
};
