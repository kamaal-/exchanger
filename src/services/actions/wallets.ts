import { WALLETS, CONVERT } from "../action_types/";
import { ThunkDispatch } from "redux-thunk";
import { saveInLocalStorage } from "../../utils/";
import Wallet from "../../interfaces_types/Wallet";
import { StoreState, Balance, BalancePair } from "../../interfaces_types/types";
import Transaction from "../models/Transaction";
import ITransaction from "../../interfaces_types/Transaction";

export const setFromWallet = (wallet: Wallet) => async (
  dispatch: ThunkDispatch<{}, {}, any>,
  getState: () => StoreState
) => {
  let { balances } = getState().Wallets;
  balances = balances ? { ...balances } : {};
  await saveInLocalStorage("fromWallet", wallet);
  dispatch({ type: WALLETS.FROM, payload: wallet });
  balances["fromWallet"] = {
    currency: wallet.currency,
    balance: wallet.balance,
    diff: 0,
    label: "Balance"
  };
  dispatch({ type: CONVERT.RUN, payload: balances });
};

export const setLeftWallet = (wallet: Wallet) => async (
  dispatch: ThunkDispatch<{}, {}, any>,
  getState: () => StoreState
) => {
  await saveInLocalStorage("leftWallet", wallet);
  dispatch({ type: WALLETS.LEFT, payload: wallet });
};

export const setToWallet = (wallet: Wallet) => async (
  dispatch: ThunkDispatch<{}, {}, any>,
  getState: () => StoreState
) => {
  let { balances } = getState().Wallets;
  balances = balances ? { ...balances } : {};
  await saveInLocalStorage("toWallet", wallet);
  dispatch({ type: WALLETS.TO, payload: wallet });
  balances["toWallet"] = {
    currency: wallet.currency,
    balance: wallet.balance,
    diff: 0,
    label: "After"
  };
  dispatch({ type: CONVERT.RUN, payload: balances });
};

export const setRightWallet = (wallet: Wallet) => async (
  dispatch: ThunkDispatch<{}, {}, any>,
  getState: () => StoreState
) => {
  await saveInLocalStorage("rightWallet", wallet);
  dispatch({ type: WALLETS.RIGHT, payload: wallet });
};

export const switcWallets = () => (
  dispatch: ThunkDispatch<{}, {}, any>,
  getState: () => StoreState
) => {
  let { fromWallet, toWallet } = getState().Wallets;
  if (fromWallet && toWallet) {
    dispatch(setToWallet(fromWallet));
    dispatch(setFromWallet(toWallet));
    dispatch(triggerConversion("0", toWallet, fromWallet));
  }
};

export const resetWallets = (fromWallet: Wallet, toWallet: Wallet) => (
  dispatch: ThunkDispatch<{}, {}, any>
) => {
  if (fromWallet && toWallet) {
    const balances = {
      fromWallet: {
        currency: fromWallet.currency,
        balance: fromWallet.balance,
        diff: 0,
        label: "Balance"
      },
      toWallet: {
        currency: toWallet.currency,
        balance: toWallet.balance,
        diff: 0,
        label: "Balance"
      }
    };

    dispatch({
      type: WALLETS.RESET,
      payload: { balances, fromWallet, toWallet }
    });
    dispatch(setValue(""));
  }
};

export const saveTransaction = (
  transaction: ITransaction,
  key = "transactions"
) => (dispatch: ThunkDispatch<{}, {}, any>) => {
  if (transaction && transaction instanceof Transaction) {
    let transactionsList: ITransaction[] = [];
    try {
      const oldTransactions = localStorage.getItem(key);
      if (oldTransactions) {
        transactionsList = JSON.parse(oldTransactions);
      }
      transactionsList = transactionsList.map(t => {
        const obj = new Transaction();
        obj.convenientConsttructor(t);
        return obj;
      });
      transactionsList.unshift(transaction);
      saveInLocalStorage(key, transactionsList);
      dispatch({ type: WALLETS.ADD_TRANSACTION, payload: transactionsList });
    } catch (error) {
      transactionsList.unshift(transaction);
      saveInLocalStorage(key, transactionsList);
      dispatch({ type: WALLETS.ADD_TRANSACTION, payload: transactionsList });
    }
  }
};

export const triggerConversion = (
  input: string,
  fromWallet: Wallet,
  toWallet: Wallet
) => (dispatch: ThunkDispatch<{}, {}, any>) => {
  if (input && fromWallet && toWallet) {
    dispatch({ type: CONVERT.DO, payload: { input, fromWallet, toWallet } });
  }
};

export const getTransaction = (
  fromWallet: Wallet,
  toWallet: Wallet,
  balances: BalancePair
) => new Transaction(fromWallet, toWallet, balances);

export const exchange = (cb = () => {}) => (
  dispatch: ThunkDispatch<{}, {}, any>,
  getState: () => StoreState
) => {
  const state: StoreState = getState();
  const { fromWallet, toWallet, balances, wallets } = state.Wallets;

  // detuct from wallet
  if (fromWallet && toWallet && balances) {
    const fromBalance: Balance = balances["fromWallet"];
    const toBalance = balances["toWallet"];
    // only pass the exchange if their any values and value can deduct from purse.
    if (
      fromBalance.diff &&
      toBalance.diff &&
      Math.abs(fromBalance.diff) <= fromWallet.balance
    ) {
      // keep the transaction before update wallets
      const transaction = getTransaction(fromWallet, toWallet, balances);

      fromWallet.balance = fromBalance.balance;
      toWallet.balance = toBalance.balance;
      const newWallets = (wallets || [])
        .map(w => (w.currency === fromWallet.currency ? fromWallet : w))
        .map(w => (w.currency === toWallet.currency ? toWallet : w));
      dispatch(updateWallets(newWallets));
      dispatch(setFromWallet(fromWallet));
      dispatch(setToWallet(toWallet));
      const newBalances = {
        fromWallet: {
          currency: fromWallet.currency,
          balance: fromWallet.balance,
          diff: 0,
          label: "Balance"
        },
        toWallet: {
          currency: toWallet.currency,
          balance: toWallet.balance,
          diff: 0,
          label: "Balance"
        }
      };
      dispatch({
        type: WALLETS.RESET,
        payload: { balances: newBalances, fromWallet, toWallet }
      });
      dispatch(saveTransaction(transaction));
      dispatch(resetWallets(fromWallet, toWallet));
      dispatch(setValue(""));
    }
  }
};

export const updateWallets = (wallets: Wallet[]) => async (
  dispatch: ThunkDispatch<{}, {}, any>
) => {
  if (wallets && wallets.length >= 0) {
    await saveInLocalStorage("wallets", wallets);
    dispatch({ type: WALLETS.ADD, payload: wallets });
  }
};

export const setValue = (value: string) => async (
  dispatch: ThunkDispatch<{}, {}, any>
) => {
  dispatch({ type: WALLETS.SET_VALUE, payload: value });
};
