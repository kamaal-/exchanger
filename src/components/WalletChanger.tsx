import React, {SFC} from 'react'
import {connect} from 'react-redux'
import { StoreState, WalletSides, Size } from '../interfaces_types/types'
import IWallet from '../interfaces_types/Wallet'
import {CircleButton, CurrencyLogo, WalletColumn} from '../theme/'
import {setFromWallet, setToWallet, setLeftWallet, setRightWallet, resetWallets} from '../services/actions/wallets'

type OwnProps = {
    wallets: IWallet[];
    toWallet: IWallet;
    fromWallet: IWallet;
    leftWallet: IWallet;
    rightWallet: IWallet;
    side: WalletSides;
}

type StateProps = {
    
}

type DispatchProps = {
    setFromWallet: (wallet: IWallet) => void;
    setToWallet: (wallet: IWallet) => void;
    setLeftWallet: (wallet: IWallet) => void;
    setRightWallet: (wallet: IWallet) => void;
    resetWallets: (fromWallet:IWallet, toWallet: IWallet) => void;
}

const mapStateToProps = (state: StoreState):StateProps => ({
    
});

const dispatchProps = {
    setFromWallet, 
    setToWallet, 
    setLeftWallet, 
    setRightWallet,
    resetWallets
};

type IProps = StateProps & OwnProps & DispatchProps;

const WalletChanger:SFC<IProps> = ({
    wallets, 
    side, 
    fromWallet,
    toWallet,
    leftWallet,
    rightWallet,
    setFromWallet, 
    setToWallet, 
    setLeftWallet, 
    resetWallets,
    setRightWallet}) => {

    const leftWalletClick = (w:IWallet) => {
        setLeftWallet && setLeftWallet(w);
        if(fromWallet.currency === leftWallet.currency){
            resetWallets && resetWallets(w, toWallet)
        }else{
            resetWallets && resetWallets(toWallet, w)
        }
    }

    const rightWalletClick = (w:IWallet) => {
        setRightWallet && setRightWallet(w);
        if(fromWallet.currency === rightWallet.currency){
            resetWallets && resetWallets(w, toWallet)
        }else{
            resetWallets && resetWallets(toWallet, w)
        }
    }
    return (
        <WalletColumn style={{width: '45px', margin: '0px'}}>
            {(wallets||[]).map((w, i) => (
                <CircleButton key={`wc-${i}-${side}`} onClick={() => {
                    if(side === WalletSides.Left){
                        leftWalletClick(w)
                    }else{
                        rightWalletClick(w)
                    }
                }}>
                    <CurrencyLogo data-testid={`walletchanger-logo-${i}-${side}`} size={Size.Small} currency={w.currency}>{w.symbol}</CurrencyLogo>
                </CircleButton>
            ))}
        </WalletColumn>
    );
}

export default connect(mapStateToProps, dispatchProps)(WalletChanger)