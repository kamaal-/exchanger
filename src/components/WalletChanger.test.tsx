import React from 'react'
import { createStore } from 'redux'
import { Provider } from 'react-redux'
import WalletChanger from './WalletChanger'
import { WalletSides } from '../interfaces_types/types'
import { render } from '@testing-library/react'
import {balances, wallets} from '../__mocks__/wallets'

const wallet = wallets[0];

describe('Input component', () => {
    function renderWithRedux(
        ui,
        { _, store = createStore(() => ({Wallets:{balances}})) } = {}
      ) {
        return {
          ...render(<Provider store={store}>{ui}</Provider>),
          // adding `store` to the returned utilities to allow us
          // to reference it in our tests (just try to avoid using
          // this to test implementation details).
          store,
        }
    }

    const setup = () => {
        return renderWithRedux(<WalletChanger toWallet={wallets[1]} fromWallet={wallet} leftWallet={wallet} rightWallet={wallets[1]} wallets={wallets} side={WalletSides.Left}/>)
    }

    it('should match snapshot', () => {
        const { asFragment } = setup();
        expect(asFragment()).toMatchSnapshot();
    });

    it('should render with logo', () => {
        const {getByTestId} = setup();
        expect.assertions(3);
        expect(getByTestId(`walletchanger-logo-0-leftWallet`).textContent).toBe('â‚¬');
        expect(getByTestId(`walletchanger-logo-1-leftWallet`).textContent).toBe('$');
        expect(getByTestId(`walletchanger-logo-2-leftWallet`).textContent).toBe('l');
    });

});