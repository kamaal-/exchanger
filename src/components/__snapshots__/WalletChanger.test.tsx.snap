// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input component should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-iwsKbI cKSZhW"
    style="width: 45px; margin: 0px;"
  >
    <div
      class="sc-htoDjs epMEqR"
    >
      <div
        class="sc-bZQynM cZOXEw"
        data-testid="walletchanger-logo-0-leftWallet"
      >
        â‚¬
      </div>
    </div>
    <div
      class="sc-htoDjs epMEqR"
    >
      <div
        class="sc-bZQynM ktwGPa"
        data-testid="walletchanger-logo-1-leftWallet"
      >
        $
      </div>
    </div>
    <div
      class="sc-htoDjs epMEqR"
    >
      <div
        class="sc-bZQynM dmSDbR"
        data-testid="walletchanger-logo-2-leftWallet"
      >
        l
      </div>
    </div>
  </div>
</DocumentFragment>
`;
